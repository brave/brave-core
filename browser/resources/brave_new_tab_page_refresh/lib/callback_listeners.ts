/* Copyright (c) 2025 The Brave Authors. All rights reserved.
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this file,
 * You can obtain one at https://mozilla.org/MPL/2.0/. */

type Router<T> = {
  [P in keyof T]: {
    addListener: (listener: any) => void
    removeListener: (listener: any) => void
  }
}

export function addCallbackListeners<T>(
  router: Router<T>,
  listeners: Partial<T>
) {
  for (const [key, value] of Object.entries(listeners)) {
    router[key as keyof T].addListener(value)
  }
  return () => {
    for (const [key, value] of Object.entries(listeners)) {
      router[key as keyof T].removeListener(value)
    }
  }
}
