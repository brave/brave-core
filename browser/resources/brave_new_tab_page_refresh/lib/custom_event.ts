/* Copyright (c) 2025 The Brave Authors. All rights reserved.
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this file,
 * You can obtain one at https://mozilla.org/MPL/2.0/. */

export type CustomEventName = 'ntp-open-news-feed-settings'

export function dispatchCustomEvent(name: CustomEventName) {
  window.dispatchEvent(new CustomEvent(name))
}

export function addCustomEventListener(
  name: CustomEventName,
  listener: () => void
) {
  window.addEventListener(name, listener)
  return () => window.removeEventListener(name, listener)
}
