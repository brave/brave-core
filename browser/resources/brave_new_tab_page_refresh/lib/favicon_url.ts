/* Copyright (c) 2025 The Brave Authors. All rights reserved.
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this file,
 * You can obtain one at https://mozilla.org/MPL/2.0/. */

export function faviconURL(url: string) {
  return 'chrome://favicon2/?size=64&pageUrl=' + encodeURIComponent(url)
}
