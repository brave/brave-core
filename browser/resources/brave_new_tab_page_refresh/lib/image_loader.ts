/* Copyright (c) 2025 The Brave Authors. All rights reserved.
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this file,
 * You can obtain one at https://mozilla.org/MPL/2.0/. */

export const placeholderImageSrc =
  'data:image/svg+xml,%3Csvg xmlns="http://www.w3.org/2000/svg"%3E%3C/svg%3E'

// Loads an image in the background and resolves when the image has either
// loaded or was unable to load.
export function loadImage(url: string): Promise<boolean> {
  return new Promise((resolve) => {
    if (!url) {
      resolve(false)
      return
    }

    const unlisten = () => {
      image.removeEventListener('load', onLoad)
      image.removeEventListener('error', onError)
    }

    const onLoad = () => {
      unlisten()
      resolve(true)
    }

    const onError = () => {
      unlisten()
      resolve(false)
    }

    const image = new Image()
    image.addEventListener('load', onLoad)
    image.addEventListener('error', onError)
    image.src = url
  })
}
