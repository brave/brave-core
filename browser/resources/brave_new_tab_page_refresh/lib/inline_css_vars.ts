/* Copyright (c) 2025 The Brave Authors. All rights reserved.
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this file,
 * You can obtain one at https://mozilla.org/MPL/2.0/. */

interface CSSVars {
  [key: `--${string}`]: string | number
}

// Allows a collection of CSS custom variables to be used in the "style" prop of
// React components.
export function inlineCSSVars(vars: CSSVars) {
  return vars as React.CSSProperties
}
