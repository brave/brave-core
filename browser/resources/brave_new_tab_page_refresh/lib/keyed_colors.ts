/* Copyright (c) 2025 The Brave Authors. All rights reserved.
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this file,
 * You can obtain one at https://mozilla.org/MPL/2.0/. */

function hashString(str: string) {
  let hash = 0
  for (let i = 0; i < str.length; ++i) {
    hash = Math.imul(31, hash) + str.charCodeAt(i)
  }
  return (hash | 0) + 2147483647 + 1
}

const colors = [
  '#FF9AA2',
  '#FFB7B2',
  '#FFDAC1',
  '#E2F0CB',
  '#B5EAD7',
  '#C7CEEA'
]

export function getKeyedColor(key: string) {
  return colors[hashString(key) % colors.length]
}
