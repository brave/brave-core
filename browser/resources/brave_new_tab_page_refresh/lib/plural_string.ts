/* Copyright (c) 2025 The Brave Authors. All rights reserved.
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this file,
 * You can obtain one at https://mozilla.org/MPL/2.0/. */

import * as React from 'react'
import { PluralStringProxyImpl } from 'chrome://resources/js/plural_string_proxy.js'

export function usePluralString(key: string, count: number) {
  const [pluralString, setPluralString] = React.useState('')
  React.useEffect(() => {
    PluralStringProxyImpl
      .getInstance()
      .getPluralString(key, count)
      .then(setPluralString)
  }, [key, count])
  return pluralString
}
