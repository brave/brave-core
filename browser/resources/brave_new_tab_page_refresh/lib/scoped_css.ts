/* Copyright (c) 2025 The Brave Authors. All rights reserved.
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this file,
 * You can obtain one at https://mozilla.org/MPL/2.0/. */

// Adds CSS to the document. If a stylesheet with the specified `id` has already
// been added to the document, then it will be replaced with the provided CSS.
async function addStyles(cssText: unknown) {
  const stylesheet = new CSSStyleSheet()
  document.adoptedStyleSheets.push(stylesheet)
  await stylesheet.replace(String(cssText))
}

const scopeAttributeName = 'data-css-scope'

class ScopedCSSAttribute {
  [scopeAttributeName]: string

  constructor(scopeName: string) {
    this[scopeAttributeName] = scopeName
  }

  get scope() {
    return this[scopeAttributeName]
  }

  get selector() {
    return `[${scopeAttributeName}=${CSS.escape(this[scopeAttributeName])}]`
  }
}

let nextScopeID = 0x5c09ed;

// A template tag that adds scoped CSS to the document. The provided CSS text
// is wrapped with a "@scope" at-rule and only applies to elements with a
// "data-css-scope" attribute whose value matches `scopeName`. The CSS rules do
// not apply to any descendant elements that have a "data-css-scope" attribute.
// Returns an object representing the CSS scope data attribute, which can be
// object-spread into a collection of HTML attributes.
export const scoped = {
  css(callsite: TemplateStringsArray, ...values: any[]) {
    const id = (nextScopeID++).toString(36)
    const attr = new ScopedCSSAttribute(id)
    addStyles(`
      @scope (${attr.selector}) to ([${scopeAttributeName}]) {
        ${String.raw(callsite, ...values)}
      }
    `)
    return attr
  }
}

// Adds global CSS to the document.
export const global = {
  css(callsite: TemplateStringsArray, ...values: any[]) {
    addStyles(String.raw(callsite, ...values))
  }
}
