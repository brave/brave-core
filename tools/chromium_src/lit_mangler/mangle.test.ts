// Copyright (c) 2025 The Brave Authors. All rights reserved.
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this file,
// You can obtain one at https://mozilla.org/MPL/2.0/.

import fs from 'fs'
import path from 'path'
import * as diff from 'diff'
import {genPath, outputPath} from '../../../build/commands/lib/guessConfig'

// Note: Headers are stripped during preprocess, we're not interested in them
// showing up in the snapshot.
const header = `// Copyright XXXX The Chromium Authors
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

`

const root = path.resolve(__dirname, '..', '..', '..')
const chromiumSrc = path.join(root, 'chromium_src')
function* walkManglers(root=chromiumSrc) {
    for (const child of fs.readdirSync(root)) {
        const childPath = path.join(root, child)
        if (fs.statSync(childPath).isDirectory()) {
            for (const match of walkManglers(childPath)) {
                yield match
            }
        } else if (childPath.endsWith('.lit_mangler.ts')) {
            yield childPath
        }
    }
}

// Sometimes mangled files are not in a directory corresponding to the source
// file. This function tries to find the mangled file by traversing up the
// directory structure and inserting '/preprocessed/' at various positions.
// https://github.com/brave/brave-browser/issues/46817
function findMangledFileAndOriginalFile(
  manglerPath: string,
  genPath: string,
  root: string,
): {mangledPath: string, originalPath: string} {
  const baseName = path.basename(manglerPath)
  const mangledPathWithoutProcessed = path.relative(
    root,
    path.join(genPath, manglerPath),
  )
  const pathParts = path
    .relative(genPath, path.resolve(mangledPathWithoutProcessed))
    .split(path.sep)

  const assumedMangledPath = path.relative(
    root,
    path.join(
      genPath,
      manglerPath.replace(baseName, 'preprocessed/' + baseName),
    ),
  )
  let mangledPath = assumedMangledPath
  const originalPath = path.join('..', manglerPath)

  let attempts = 0
  while (attempts < pathParts.length - 1 && !fs.existsSync(mangledPath)) {
    // Try inserting 'preprocessed' at position
    // e.g.
    // iter 1) a/b/preprocessed/c/file
    // iter 2) a/preprocessed/b/c/file
    const insertPosition = Math.max(0, pathParts.length - 2 - attempts)
    const newPathParts = [
      ...pathParts.slice(0, insertPosition),
      'preprocessed',
      ...pathParts.slice(insertPosition),
    ]

    mangledPath = path.join(genPath, ...newPathParts)

    attempts++
  }
  if (!fs.existsSync(mangledPath)) {
    throw new Error(`Mangled file not found: ${assumedMangledPath}`)
  }

  return {mangledPath, originalPath}
}

describe('mangled files should have up to date snapshots', () => {
    for (const mangler of walkManglers()) {
        const name = path.relative(root, mangler).replaceAll(path.sep, '/')
        it(`./${name} should match snapshot`, () => {
            const manglerPath = name.replace('.lit_mangler.ts', '')
                .replace(`chromium_src/`, '')

            const {mangledPath, originalPath}= findMangledFileAndOriginalFile(manglerPath, genPath, root)


            const mangledText = fs.readFileSync(mangledPath, 'utf8')
                .replaceAll('\r\n', '\n')
            const originalText = fs.readFileSync(originalPath, 'utf8')
                .replaceAll('\r\n', '\n')
                .substring(header.length)
            const linesDiff = diff.createTwoFilesPatch(originalPath.replaceAll(path.sep, '/'),
                path.relative(outputPath, mangledPath).replaceAll(path.sep, '/'),
                originalText,
                mangledText)
            expect(linesDiff).toMatchSnapshot()
        });
    }
});
