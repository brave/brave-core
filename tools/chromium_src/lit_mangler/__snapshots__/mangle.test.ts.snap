// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mangled files should have up to date snapshots ./chromium_src/chrome/browser/resources/app_settings/window_mode_item.html.ts.lit_mangler.ts should match snapshot 1`] = `
"===================================================================
--- ../chrome/browser/resources/app_settings/window_mode_item.html.ts
+++ gen/chrome/browser/resources/app_settings/preprocessed/window_mode_item.html.ts
@@ -3,11 +3,16 @@
 import type {WindowModeItemElement} from './window_mode_item.js';
 
 export function getHtml(this: WindowModeItemElement) {
   return html\`<!--_html_template_start_-->
-<app-management-toggle-row
-    id=\\"toggle-row\\"
-    label=\\"\${this.windowModeLabel}\\"
-    ?value=\\"\${this.getValue_()}\\">
-</app-management-toggle-row>
+<div ?hidden=\\"\${this.app.hideWindowMode}\\">
+    <div class=\\"permission-section-header\\">
+      <div class=\\"header-text\\">$i18n{appManagementOpenModeLabel}</div>
+    </div>
+    <div class=\\"permission-list indented-permission-block\\">
+      \${this.windowModes.map(mode => html\`<leo-radiobutton class=\\"subpermission-row\\" name=\\"window-mode\\" @change=\\"\${this.onChange}\\" value=\\"\${mode.value}\\" currentvalue=\\"\${this.app.windowMode}\\">
+        \${mode.label}
+      </leo-radiobutton>\`)}
+    </div>
+  </div>
 <!--_html_template_end_-->\`;
 }
"
`;

exports[`mangled files should have up to date snapshots ./chromium_src/chrome/browser/resources/extensions/detail_view.html.ts.lit_mangler.ts should match snapshot 1`] = `
"===================================================================
--- ../chrome/browser/resources/extensions/detail_view.html.ts
+++ gen/chrome/browser/resources/extensions/preprocessed/detail_view.html.ts
@@ -351,16 +351,13 @@
               @change=\\"\${this.onPinnedToToolbarChange_}\\">
             <span>$i18n{itemPinToToolbar}</span>
           </extensions-toggle-row>\` : ''}
         \${this.shouldShowIncognitoOption_() ? html\`
-          <extensions-toggle-row id=\\"allow-incognito\\"
-              ?checked=\\"\${this.data.incognitoAccess.isActive}\\" class=\\"hr\\"
-              @change=\\"\${this.onAllowIncognitoChange_}\\">
+          <extensions-toggle-row id=\\"allow-incognito\\" ?checked=\\"\${this.data.incognitoAccess.isActive}\\" class=\\"hr\\" @change=\\"\${this.onAllowIncognitoChange_}\\">
             <div>
               <div>$i18n{itemAllowIncognito}</div>
-              <div class=\\"section-content\\">$i18n{incognitoInfoWarning}</div>
-              <div class=\\"section-content\\" id=\\"allow-incognito-warning\\"
-                  ?hidden=\\"\${!this.data.incognitoAccessPendingChange}\\">
+              <div class=\\"section-content\\">$i18n{privateInfoWarning}<span ?hidden=\\"\${!this.data.incognitoAccess.isActive}\\"> $i18n{spanningInfoWarning}</span><span> $i18n{privateAndTorInfoWarning}</span></div>
+              <div class=\\"section-content\\" id=\\"allow-incognito-warning\\" ?hidden=\\"\${!this.data.incognitoAccessPendingChange}\\">
                 $i18n{pendingChangeWarning}
               </div>
             </div>
           </extensions-toggle-row>\` : ''}
"
`;
